include/tvm/relay/expr.h

```diff
@@ -0,0 +1,373 @@
+  TVM_DECLARE_NODE_TYPE_INFO(LocalVarNode, ExprNode);
+};
+
+RELAY_DEFINE_NODE_REF(LocalVar, LocalVarNode, Expr);
+
+/*!
+ * \brief Global variable that leaves in the top-level environment.
+ * This is used to enable recursive calls between function.
+ *
+ * \note GlobalVar can only corresponds to functions.
+ */
+class GlobalVar;
+/*! \brief A GlobalId from the node's current type to target type. */
+class GlobalVarNode : public ExprNode {
+ public:
+  /*! \brief The name of the variable, this only acts as a hint. */
+  std::string name_hint;
+
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+    v->Visit("name_hint", &name_hint);
+  }
+
+  TVM_DLL static GlobalVar make(std::string name_hint);
+
+  static constexpr const char* _type_key = "relay.GlobalVar";
+  TVM_DECLARE_NODE_TYPE_INFO(GlobalVarNode, ExprNode);
+};
+
+RELAY_DEFINE_NODE_REF(GlobalVar, GlobalVarNode, Expr);
+
+/*!
+ * \brief Function parameter declaration.
+ */
+class Param;
+/*! \brief A parameter. */
+class ParamNode : public ExprNode {
+ public:
+  /*! \brief The variable */
+  LocalVar var;
+  /*! \brief The type of the parameter */
+  Type type;
+
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+    v->Visit("var", &var);
+    v->Visit("type", &type);
+    v->Visit("span", &span);
+  }
+
+  TVM_DLL static Param make(LocalVar var, Type type);
+
+  static constexpr const char* _type_key = "relay.Param";
+  TVM_DECLARE_NODE_TYPE_INFO(ParamNode, ExprNode);
+};
+
+RELAY_DEFINE_NODE_REF(Param, ParamNode, Expr);
+
+/*!
+ * \brief Function (subgraph in computational graph)
+ */
+class Function;
+/*! \brief Function container */
+class FunctionNode : public ExprNode {
+ public:
+  /*! \brief Function parameters */
+  tvm::Array<Param> params;
+  /*! \brief User annotated return type of the function. */
+  Type ret_type;
+  /*!
+   * \brief
+   * The expression which represents the computation of the function,
+   * the expression may reference the parameters, and the type of it
+   * or sub-expressions may reference the type variables.
+   */
+  Expr body;
+  /*!
+   * \brief Type parameters of the function.
+   *  Enables the function to vary its type based on these.
+   *  This corresponds to template paramaters in c++'s terminology.
+   *
+   * \note This can be usually empty for non-polymorphic functions.
+   */
+  tvm::Array<TypeParam> type_params;
+
+  void VisitAttrs(tvm::AttrVisitor* v) final {
+    v->Visit("params", &params);
+    v->Visit("ret_type", &ret_type);
+    v->Visit("body", &body);
+    v->Visit("type_params", &type_params);
+    v->Visit("span", &span);
+  }
+
+  TVM_DLL static Function make(tvm::Array<Param> params, Type ret_type,
+                               Expr body, tvm::Array<TypeParam> ty_params);
+
+  static constexpr const char* _type_key = "relay.Function";
+  TVM_DECLARE_NODE_TYPE_INFO(FunctionNode, ExprNode);
+};
+
+RELAY_DEFINE_NODE_REF(Function, FunctionNode, Expr);
+
+using Attrs = tvm::Attrs;
```

Comment:
nit: Move it into class CallNode? Seems that's the only place you need it.

Reply:
Yeah I think this is an old artifact of when this was typedef'd to something else. I will just remove and use `tvm::Attrs`.

Reply:
I would also prefer tvm::Attrs